libclc/amdgcn-amdhsa/libspirv/misc/sub_group_shuffle.cl

```diff
@@ -256,12 +256,26 @@ __AMDGCN_CLC_SUBGROUP_XOR_TO_VEC(double16, d, 16)
 // Shuffle Up
 // int __spirv_SubgroupShuffleUpINTEL<int>(int, int, unsigned int)
 _CLC_DEF int
-_Z30__spirv_SubgroupShuffleUpINTELIiET_S0_S0_j(int var, int lane_delta,
-                                               unsigned int width) {
+_Z30__spirv_SubgroupShuffleUpINTELIiET_S0_S0_j(int previous, int current,
+                                               unsigned int delta) {
   int self = SELF;
-  int index = self - lane_delta;
-  index = (index < (self & ~(width - 1))) ? index : self;
-  return __builtin_amdgcn_ds_bpermute(index << 2, var);
+  int size = SUBGROUP_SIZE;
+
+  int index = self - delta;
+
+  int val;
+  if (index >= 0 && index < size) {
+    val = current;
+  } else if (index < 0 && index > -size) {
+    val = previous;
+    index = index + size;
+  } else {
+    // index out of bounds so return arbitrary data
+    val = current;
+    index = self;
+  }
```

Comment:
This looks like a correct implementation of `ShuffleUpINTEL` to me, but I'm concerned about the extra complexity it introduces.  We know that `previous` and `current` are always the same variable (if this intrinsic is called from the SYCL headers.

The SPIR-V non-uniform shuffle instructions (e.g. [OpGroupNonUniformShuffleUp](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformShuffleUp)) are much closer semantically to the functionality required by SYCL, and this PR has me wondering whether it would be a good idea to switch over to those.  It might be worth opening an issue or something to track this -- somebody would have to run some experiments to see which backends actually support these instructions, etc.

Reply:
Agreed, we should definitely investigate the performance of this at some point.

Also for what it's worth the CUDA backend implements this directly in the header using NVPTX built-ins rather than going through the SPIR-V interface. So that mostly would affect the HIP backend and maybe others as well.

I've filed a ticket to keep track of this:
* https://github.com/intel/llvm/issues/5364